lib/flutter/flutter_daemon.dart

```diff
@@ -0,0 +1,373 @@
+
+  Stream<String> get onSend => _onSend.stream;
+  Stream<String> get onReceive => _onReceive.stream;
+
+  set willSend(void fn(String methodName)) {
+    _willSend = fn;
+  }
+
+  void dispose() {
+    if (_streamSub != null) _streamSub.cancel();
+    //_completers.values.forEach((c) => c.completeError('disposed'));
+    _completers.clear();
+    otherDisposeable?.dispose();
+  }
+
+  void _processMessage(String message) {
+    try {
+      _onReceive.add(message);
+
+      var json = JSON.decode(message);
+
+      if (json['id'] == null) {
+        // Handle a notification.
+        String event = json['event'];
+        String prefix = event.substring(0, event.indexOf('.'));
+        if (_domains[prefix] == null) {
+          _logger.severe('no domain for notification: ${message}');
+        } else {
+          _domains[prefix]._handleEvent(event, json['params']);
+        }
+      } else {
+        Completer completer = _completers.remove(json['id']);
+        String methodName = _methodNames.remove(json['id']);
+
+        if (completer == null) {
+          _logger.severe('unmatched request response: ${message}');
+        } else if (json['error'] != null) {
+          completer.completeError(new RequestError(methodName, json['error']));
+        } else {
+          completer.complete(json['result']);
+        }
+      }
+    } catch (e) {
+      _logger.severe('unable to decode message: ${message}, ${e}');
+    }
+  }
+
+  Future<dynamic> _call(String method, [Map args]) {
+    String id = '${++_id}';
+    _completers[id] = new Completer();
+    _methodNames[id] = method;
+    Map m = {'id': id, 'method': method};
+    if (args != null) m['params'] = args;
+    String message = _jsonEncoder.encode(m);
+    if (_willSend != null) _willSend(method);
+    _onSend.add(message);
+    _writeMessage(message);
+    return _completers[id].future;
+  }
+
+  static dynamic _toEncodable(obj) => obj is Jsonable ? obj.toMap() : obj;
+}
+
+abstract class Domain {
+  final FlutterDaemon server;
+  final String name;
+
+  Map<String, StreamController> _controllers = {};
+  Map<String, Stream> _streams = {};
+
+  Domain(this.server, this.name) {
+    server._domains[name] = this;
+  }
+
+  Future<dynamic> _call(String method, [Map args]) => server._call(method, args);
+
+  Stream<dynamic> _listen(String name, Function cvt) {
+    if (_streams[name] == null) {
+      _controllers[name] = new StreamController.broadcast();
+      _streams[name] = _controllers[name].stream.map(cvt);
+    }
+
+    return _streams[name];
+  }
+
+  void _handleEvent(String name, dynamic event) {
+    if (_controllers[name] != null) {
+      _controllers[name].add(event);
+    }
+  }
+
+  String toString() => 'Domain ${name}';
+}
+
+abstract class Jsonable {
+  Map toMap();
+}
+
+class RequestError {
+  final String methodName;
+  final dynamic error;
+
+  RequestError(this.methodName, this.error);
+
+  String toString() => '[RequestError ${error}]';
+}
+
+Map _mapify(Map m) {
```

Comment:
Would it make sense to replace calls to this method with calls to Map.from(...) or does this method do something different that's important to its callers?